feat: esm migration - Adopt Pure ESM Build and Refresh Test Harness for Vite 7 and Vitest 3

hmrkx opened the pull request:
- Raise the SDK baseline to Node 20+ and publish a single bundled ESM entry (dist/index.mjs) with embargoed CJS exports
  - Upgrade runtime/tooling deps (ethereumjs, viem, vitest/vite, typescript, eslint, etc.) to their latest majors and align tsup/
  tsconfig with ES2022 bundler resolution
  - Harden lint rules against bare core-module imports and CommonJS fallbacks
  - Split test bootstrap (testEnvironment.ts) from reusable client-storage helpers, streamline them for Node 20, and ensure dotenv
  - global matchers load once via Vitest setup
  - validate the new stack with pnpm build, pnpm exec tsc --noEmit, and the full unit suite on Vite 7/Vitest 3

hmrkx committed:
feat: Adopt Pure ESM Build and Refresh Test Harness for Vite 7/Vitest 3

yilmazbahadir committed:
fix: correct import path in test utilities

yilmazbahadir commented:
### Test Plan
1. Run the [lattice-simulator](https://github.com/GridPlus/lattice-simulator):
`LATTICE_AUTO_APPROVE=true DEBUG_SIGNING=1 pnpm dev`

3. 
```
DEBUG_EVM_SIGN=1 baseUrl=http://127.0.0.1:3000 DEVICE_ID=SD0001 PASSWORD=12345678 PAIRING_SECRET=12345678 APP_NAME=lattice-manager pnpm e2e-sign-evm-tx -- --reporter=basic
```

 ✓ src/__test__/e2e/signing/evm-tx.test.ts (36 tests) 47955ms
   ✓ EVM Transaction Signing - Unified Test Suite > Legacy Transactions > Simple ETH transfer - Legacy (1/5)  1415ms
   ✓ EVM Transaction Signing - Unified Test Suite > Legacy Transactions > Contract interaction - Legacy (2/5)  2364ms
   ✓ EVM Transaction Signing - Unified Test Suite > Legacy Transactions > Zero value transaction - Legacy (3/5)  1166ms
   ✓ EVM Transaction Signing - Unified Test Suite > Legacy Transactions > High nonce transaction - Legacy (4/5)  1047ms
   ✓ EVM Transaction Signing - Unified Test Suite > Legacy Transactions > Polygon Legacy transaction (5/5)  1275ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-1559 Transactions (Fee Market) > Simple ETH transfer - EIP-1559 (1/6)  1142ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-1559 Transactions (Fee Market) > DAI transfer - EIP-1559 (2/6)  2080ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-1559 Transactions (Fee Market) > Uniswap V2 swap - EIP-1559 (3/6)  1488ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-1559 Transactions (Fee Market) > WETH deposit - EIP-1559 (4/6)  1419ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-1559 Transactions (Fee Market) > High priority fee - EIP-1559 (5/6)  1020ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-1559 Transactions (Fee Market) > Polygon EIP-1559 transaction (6/6)  994ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-2930 Transactions (Access Lists) > Simple transfer with access list - EIP-2930 (1/4)  1189ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-2930 Transactions (Access Lists) > Contract interaction with multiple access entries - EIP-2930 (2/4)  1831ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-2930 Transactions (Access Lists) > Empty access list - EIP-2930 (3/4)  1142ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-2930 Transactions (Access Lists) > Complex DeFi interaction - EIP-2930 (4/4)  1459ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-7702 Transactions (Account Abstraction) > Real mainnet EIP-7702 transaction - Simple authorization (1/3)  1269ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-7702 Transactions (Account Abstraction) > Real mainnet EIP-7702 transaction - Different authorization (2/3)  1462ms
   ✓ EVM Transaction Signing - Unified Test Suite > EIP-7702 Transactions (Account Abstraction) > Real mainnet EIP-7702 transaction - High value authorization (3/3)  1299ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Maximum nonce value - Legacy (1/18)  1053ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Maximum gas price - Legacy (2/18)  1084ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Minimal gas limit - EIP-1559 (3/18)  1047ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Large data payload - EIP-1559 (4/18)  1803ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Contract creation - Legacy (5/18)  1479ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Zero gas price - Legacy (pre-EIP-155) (6/18)  1128ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Alternative chain IDs - Polygon (7/18)  1012ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Alternative chain IDs - BSC (8/18)  1264ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Alternative chain IDs - Avalanche (9/18)  1094ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Large chain ID - Palm Network (10/18)  1179ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Unknown chain ID (11/18)  1221ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Minimal value - 1 wei (12/18)  1047ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Scientific notation value - 1e8 wei (13/18)  1273ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Maximum safe value (14/18)  1248ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Contract creation - EIP-1559 (15/18)  1888ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > Contract creation - EIP-2930 (16/18)  1481ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > EIP-2930 with mixed storage keys (17/18)  979ms
   ✓ EVM Transaction Signing - Unified Test Suite > Edge Cases & Boundary Conditions > EIP-1559 with access list (18/18)  1148ms

 Test Files  1 passed (1)
      Tests  36 passed (36)
   Start at  11:45:15
   Duration  49.80s (transform 246ms, setup 17ms, collect 1.13s, tests 47.95s, environment 0ms, prepare 49ms)

hmrkx committed:
feat: update imports

hmrkx committed:
feat: updated branch

hmrkx committed:
feat: update lock file

hmrkx committed:
fix: eth bip32 test

hmrkx committed:
Merge branch 'dev' into feat/esm-migration

hmrkx committed:
feat: improve package versions

hmrkx committed:
feat: no lint errors

hmrkx committed:
feat: added dev dependencies

hmrkx committed:
feat: fixing test errors

hmrkx committed:
feat: fixing errors

hmrkx committed:
feat: btc tests

hmrkx committed:
feat: sign

hmrkx committed:
feat: segwit tests

hmrkx committed:
feat: fix sign

netbonus commented on src/__test__/utils/runners.ts:
```diff
@@ -57,26 +55,22 @@ export async function runEvm(
   // Request signature and validate it
   await client.connect(getDeviceId());
-  const resp = await client.sign(req);
+  const resp: any = await client.sign(req);
```
this should have a type

netbonus commented on vite.config.ts:
```diff
@@ -1,49 +1,18 @@
-import { resolve } from 'node:path';
 
 export default defineConfig({
-  build: {
```
is our build output the same as it was before? what's different?

hmrkx replied:
The main differences would now be:
  - Build moved from Vite to tsup.
      - Old: Vite library build with Rollup options and vite-plugin-dts.
      - New: tsup-only build, ESM output, types emitted by tsup.
  - Vite config now only configures Vitest.
  - Package is ESM-only ("type": "module"), with exports pointing to the ESM bundle.

As for the outputs:
  - Old outputs
      - Formats: es + umd.
      - Files: gridplus-sdk.es.js and gridplus-sdk.umd.js (with name: 'GridPlusSDK' and UMD globals mappings).
      - DTS via vite-plugin-dts.
  - New outputs (tsup)
      - Format: esm only (Node 20 target), file: dist/index.mjs, sourcemaps on, DTS at dist/index.d.ts.
      - Externals: all dependencies and peerDependencies are excluded from the bundle (imported at runtime).
          - Old config only externalized a subset, now it externalizes everything in deps/peerDeps.

netbonus commented on src/__test__/utils/viemComparison.ts:
```diff
@@ -35,22 +27,7 @@ export const getFoundryAccount = () => {
-    transport: http(),
-  });
-};
+// (Wallet client helper removed as unused)
```
```suggestion

```

netbonus commented on src/__test__/utils/viemComparison.ts:
```diff
@@ -1,15 +1,7 @@
+import { parseTransaction, serializeTransaction, type Address, type Hex, type TransactionSerializable } from 'viem';
 import { privateKeyToAccount } from 'viem/accounts';
-import { mainnet, polygon, bsc, avalanche } from 'viem/chains';
+// Chains & wallet client not needed currently; compare via account signing.
```
```suggestion
```

netbonus commented on src/__test__/utils/setup.ts:
why remove this and refactor here?

netbonus commented on src/__test__/utils/runners.ts:
```diff
@@ -26,25 +27,22 @@ export async function runTestCase(
   const response = await client.sign(request);
-  // If no encoding type is specified we encode in hex or ascii
-  const encodingType = request.data.encodingType || null;
+  const data = request.data as unknown as MaybeGenericData;
```
we should just get the real type here, instead of casting

netbonus commented on package.json:
spacing shouldn't be changed

netbonus commented on src/__test__/e2e/non-exportable.test.ts:
do we need all of these whitespace changes?

hmrkx committed:
feat: revert setup and run prettier

hmrkx committed:
feat: types

hmrkx committed:
feat: sanitize commas

hmrkx committed:
feat: better typing

netbonus committed:
merge: integrate origin/dev into feat/esm-migration

Resolve dependency version conflicts and update lock file
- Update @types/lodash to ^4.17.20 and @types/node to ^24.5.2
- Regenerate pnpm-lock.yaml with all latest dependencies

Align test imports with ESM patterns
- Use vitest globals for mocking (vi.mock)
- Consolidate duplicate imports in test utilities
- Enforce node: prefix for core modules (fs, path, url)
- Remove unused imports and variables

All builds, type checks, and lints pass with updated toolchain

netbonus committed:
feat: add CommonJS build alongside ESM

Restore dual-format build output for broader compatibility
- Update tsup to generate both .mjs (ESM) and .cjs (CommonJS) bundles
- Add main/module fields to package.json for legacy tooling
- Update exports map to support both import and require
- Generate format-specific type declarations (.d.ts and .d.cts)

Both formats verified working:
- CJS: dist/index.cjs (174K)
- ESM: dist/index.mjs (168K)
- Build passes, lint clean, both imports functional

hmrkx reviewed (commented):
👍  `[feat: add CommonJS build alongside ESM](https://github.com/GridPlus/gridplus-sdk/pull/614/commits/a45ae2f74a0ef8084a9f2fe1d0edd868b466e753)`

netbonus committed:
merge: sync latest dependabot updates from origin/dev

Incorporate dependency and CI updates from origin/dev:
- ci: bump peaceiris/actions-gh-pages from v3 to v4
- deps: eslint-plugin-prettier and @solana/web3.js already at latest

No lockfile changes needed (versions already current)

netbonus committed:
merge: integrate latest dev changes into feat/esm-migration

Resolved dependency conflicts in package.json and pnpm-lock.yaml:
- Kept newer eslint (^9.36.0) and eslint-plugin-prettier (^5.5.4)
- Regenerated lock file with pnpm install
- All tests passing (150/150)
